### ISSUE TITLE
Auto-genereate layout schema based on TypeScript interfaces

### ISSUE THREAD

**Issue by framitdavid** [Altinn/app-frontend-react#832]
### Description

When we are creating new components within app-frontend-react we need to manually add the component interface to the layout schema. This task is time-consuming and it may be errors/mismatches between the schema and the component interface. This task should be automated to avoid the mentioned drawbacks. 

I'm not sure what is the best approach to generate the layout schema but I have listed 3 alternatives below:

1. Generate schema based on TypeScript Interfaces for each component
2. Generate TS Interface based on layout schema
3. Have a third-part format to define the interface and then convert it to schema and typescript.

**Comment by nkylstad** [comment]
Lots of possibilities here! I very much like the idea of 3, though at least in the short term 1. might be the most practical considering that the interfaces are already easily available and very structured. 

Wrt developer experience, 3 might also be more constraining in that there is an extra step that needs to be done when developing a new component. Typically the component is first developed and tested in app frontend, and then added to the jsonschema, so starting in that end and going with 1 might also be the best solution wrt that. But @olemartinorg will have to weigh in here.

One important aspect we have to consider is that the structure of the current `layout.v1.schema.json` file is not optimal. F.ex. a lot of components are defined with properties from the component base that are not relevant. Fex. `dataModelBindings` are expected for components like `Header` and `Paragraph`. 
If we autogenerate the JSON schema from the TS interfaces, it would probably solve that for us.

**Comment by olemartinorg** [comment]
I also gravitate towards 3 here, also because I'd prefer to do this once instead of having a stop-gap solution before that (we have that temporary solution already, by manually trying to keep the files in sync). This is a tooling and developer experience issue, so we shouldn't let that steal focus from delivering functionality to users (which is why I recommend doing this once).

With that said, the immediate problem I see with number 1 is that it might be hard to parse typescript definitions. The types disappear at runtime, so then we'd have to shop around for a typescript parser (for what i perceive is just a temporary solution).

I've done similar things before, and I looked into which libraries I used. Turns out it's [`io-ts`](https://gcanti.github.io/io-ts/) and [`io-ts-codegen`](https://gcanti.github.io/io-ts-codegen/). Their promise is that you can write code (once) to define complex types that generate both TypeScript definitions and JsonSchema, which you can then use to both publish your API as an OpenAPI specification, while also using the same definitions to validate submitted data to ensure it is valid according to your definitions, and have strong typing in your code. It might not be mature or flexible enough for our needs, though - but it worked fine for me in a green-field project a few years ago.

I agree that developer experience might suffer from alternative 3, but the way I see it we have two alternatives:
- Generate the types on the fly, and keep the artifacts in folders/files ignored by git (for example `*.autogen.ts`), meaning a new git clone will need to run a `yarn` command in order to get their IDE support going. Our build scripts, etc, would need to do the same. We would have no traceability on the code generation result, but we could write tests to at least make sure the JsonSchema is well-formed after changes to properties (or the generator itself).
- Generate the types on developer machines, and checking in the result via git. This would make cloning a repo and making unrelated changes easier, and we'd get traceability in auto-generated results. We could also have automatic tests in github workflows to check if a developer did indeed commit all auto-generated code if they made any changes to the source definition.

I'd prefer the first one, the more I think about it. For the same reason we don't expect developers to commit big blobs of build artifacts along with their changes.

**Comment by olemartinorg** [comment]
An idea from @TheTechArch: We could also generate markdown for docs.altinn.studio, so that each of our components get a page detailing every property that could/should be set.

And an idea from a [Slack thread that went off the rails](https://altinndevops.slack.com/archives/C9E57MDF1/p1682585114628259); we could use a common component definition to also generate layout files for a form that could be used to configure a layout component inside DevTools. We already make a great tool used to build forms that aim to place structured data into a JSON object structure that should validate against a JsonSchema. If we treat our (target) layout files as a data model, we could build Altinn 3 forms that can be used to configure Altinn 3 forms. 🤯 Proper re-use and dogfooding would make our product better, IMO.

These two suggestions further cement the need for another source-of-truth than JsonSchema, I believe.
